### PR TITLE
feat: implement CoreML NMS for Segment and Pose models

### PR DESCRIPTION
**Pull request by shafeeq27edu-ai**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
CoreML export now applies integrated NMS wrappers to Segment and Pose models in addition to Detect models, including their mask coefficients and keypoint outputs.

### 📊 Key Changes
- Allows `nms=True` for CoreML `segment` and `pose` tasks while continuing to disable it for unsupported tasks.
- Adds `IOSSegmentModel` to combine class scores with scaled mask coefficients for CoreML NMS.
- Adds `IOSPoseModel` to combine class scores with scaled keypoint values for CoreML NMS.
- Selects the appropriate Detect, Segment, or Pose wrapper during CoreML export.
- Removes the Windows platform assertion from the CoreML export path.

### 🎯 Purpose & Impact
CoreML Segment and Pose exports can now use the same integrated NMS workflow as Detect exports when `nms=True`; unsupported task handling remains unchanged.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @shafeeq27edu-ai, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution was received, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Reviewed the CoreML exporter, task heads, pipeline construction, CoreML backend, and segment/pose predictors. The new wrappers are selected correctly, but the exported segment/pose NMS outputs are not consumable by the existing CoreML backend, and legacy `.mlmodel` segment export has a nested output-shape failure.

<details><summary>💬 Posted 2 inline comments</summary>

- ⚠️ **HIGH** `ultralytics/utils/export/coreml.py:70` The appended extras are not preserved by Ultralytics CoreML inference: `ultralytics/nn/backends/coreml.py` currently reduces the NMS `confidence` output to `[box, max_conf, class]`, so `SegmentationPredictor` receives no `pred[:, 6:]` mask coefficients and `PosePredictor` cannot reconstruct its keypoints. Additionally, scaling these values by `1e-4` has no inverse in the consumer, so retaining them as-is would produce near-zero masks/keypoints. Update the CoreML output contract/backend to prese…
- 💡 **MEDIUM** `ultralytics/engine/exporter.py:1301` Segment heads return `(predictions, proto)` when `export=True`, so the earlier dry run makes `self.output_shape` a nested tuple. For `format=mlmodel`, `pipeline_coreml()` still assumes a flat detect shape and evaluates `output_shape[2]` and `output_shape[1] - 4`, causing `nms=True` segment exports to fail with `IndexError`. Pass the prediction tensor's shape or normalize `output_shape` before invoking the pipeline.
</details>

File: ultralytics/utils/export/coreml.py [hunk 1]
```diff
@@ -45,6 +45,58 @@ def forward(self, x: torch.Tensor):
...
+        preds = preds[0]  # strip batch dimension
+        xywh, cls, masks = preds.transpose(0, 1).split((4, self.nc, self.nm), 1)
+
+        if self.mlprogram and self.nc % 80 != 0:
+            pad_length = int(((self.nc + 79) // 80) * 80) - self.nc
+            cls = torch.nn.functional.pad(cls, (0, pad_length, 0, 0), "constant", 0)
+
+        # Scale mask coefficients to prevent them from becoming the argmax class during NMS
+        masks = masks * 1e-4
+        conf = torch.cat([cls, masks], dim=1)
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/export/coreml.py]: ⚠️ **HIGH**: The appended extras are not preserved by Ultralytics CoreML inference: `ultralytics/nn/backends/coreml.py` currently reduces the NMS `confidence` output to `[box, max_conf, class]`, so `SegmentationPredictor` receives no `pred[:, 6:]` mask coefficients and `PosePredictor` cannot reconstruct its keypoints. Additionally, scaling these values by `1e-4` has no inverse in the consumer, so retaining them as-is would produce near-zero masks/keypoints. Update the CoreML output contract/backend to preserve the selected extra columns and restore their scale (or emit them as separate outputs). The same issue applies to the analogous keypoint path at line 95.

File: ultralytics/engine/exporter.py [hunk 2]
```diff
@@ -1277,20 +1278,33 @@ def export_ncnn(self, prefix=colorstr("NCNN:")):  # noqa: B008
...
         assert TORCH_1_11, "CoreML export requires torch>=1.11"
         f = self.file.with_suffix(".mlmodel" if mlmodel else ".mlpackage")
         if f.is_dir():
             shutil.rmtree(f)
 
-        # TODO CoreML Segment and Pose model pipelining; 'nms=True' is forced off for non-detect tasks upstream
-        model = IOSDetectModel(self.model, self.im, mlprogram=not mlmodel) if self.args.nms else self.model
+        if self.args.nms:
+            if self.model.task == "segment":
+                model = IOSSegmentModel(self.model, self.im, mlprogram=not mlmodel)
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/exporter.py]: 💡 **MEDIUM**: Segment heads return `(predictions, proto)` when `export=True`, so the earlier dry run makes `self.output_shape` a nested tuple. For `format=mlmodel`, `pipeline_coreml()` still assumes a flat detect shape and evaluates `output_shape[2]` and `output_shape[1] - 4`, causing `nms=True` segment exports to fail with `IndexError`. Pass the prediction tensor's shape or normalize `output_shape` before invoking the pipeline.

**Comment by SergioAlmeida29** [comment]
Hi, I actually already have an open PR for this (#25873, fixing #25865 via NMSModel, matching ONNX/TRT/TFLite). Let's coordinate so we avoid duplicate work!

A few notes on this implementation:

- CoreML NMS issues: Native NMS treats appended masks/keypoints as class scores and filters them independently (the 1e-4 scale doesn't prevent this).

- Data dropped: The CoreML backend truncates outputs to [box, conf, class], so mask/keypoint values never reach the predictor anyway.

- Cleanup & Tests: No tests or docs are included, and the commented-out WINDOWS assert is unrelated and should probably be removed.

**Review by Y-T-G — CHANGES_REQUESTED** [review]
This variant makes the model more than 2.6 times slower, defeating the point of embedding NMS.

**without nms**

Image

**with nms**
Image

**Comment by shafeeq27edu-ai** [comment]
Thanks for pointing this out. I’ve reviewed the feedback and agree that my current approach has two important issues: the CoreML backend drops the auxiliary mask/keypoint outputs, and the implementation introduces a significant inference slowdown.

I’ll investigate the NMSModel approach used in #25873 and compare it with the existing CoreML export architecture. I’d like to avoid duplicating your work, so I’ll coordinate with you before making further changes to #25872.

**Comment by glenn-jocher** [comment]
Thanks for reviewing this. Since #25873 already covers the same change via `NMSModel` with tests and docs, please close this PR to avoid duplicate work and continue coordination there; keep the unrelated Windows assertion change separate.
